### PR TITLE
1password facts module, based on the "onepassword" lookup plugin

### PR DESCRIPTION
**Pull request by Rylon**
##### SUMMARY

As discussed with `nitzmahone` in #ansible-devel on IRC, please find attached a facts module version of the existing `onepassword.py` lookup plugin for your consideration.

The rationale for developing this version was:

- To be able to query one or more fields, from one or more items in 1Password, as a single task, returning them all as facts, which can then be used in several more places in the play. I don't think this is possible as a lookup plugin?

- To be able to implement "auto login" support, by passing credentials to the module, via Ansible Vault.  This becomes quite cumbersome and unwieldy as a lookup plugin. Note that if credentials are not set, then the existing behaviour where the user must have already run `op signin` is preserved.

- To be able to specify the location of the `op` binary, if it is not available in the default PATH.

- To be able to retrieve 1Password items which have document attachments and return their contents.

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

`1password_facts`

##### ANSIBLE VERSION

```
ansible 2.7.0.dev0 (1password-module 4cfc7c5267) last updated 2018/08/10 18:21:59 (GMT +100)
  config file = None
  configured module search path = [u'/Users/ryan/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/ryan/code/rylon-ansible/lib/ansible
  executable location = ./bin/ansible
  python version = 2.7.14 (default, Mar 27 2018, 16:13:06) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```

##### ADDITIONAL INFORMATION

- This is my first module submission and I've tried to follow the contributor guide and make sure the module is compliant, but some of my documentation and other metadata may not be quite right. If so, sorry for that, I will correct it!

- I'm not sure if the module name is acceptable, starting with a number?

- I wasn't sure which folder to put it in, so sought advice on IRC, please advise if this is incorrect.

### CODE REVIEW

**Comment by Rylon** [comment]
I will take a look at these failed tests and update the PR 👍

**Comment by Rylon** [comment]
I've pushed some fixes and CI is currently re-running, however it appears to have become stuck in the 'WAITING' state, for the past 1 hour, 22 minutes at least.

ready_for_review

Update: Never mind, there was a large queue in Shippable, seems to be processing now :)

File: lib/ansible/modules/identity/1password_facts.py [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+module: 1password_facts
```
Comment (samdoran) [inline lib/ansible/modules/identity/1password_facts.py]: This should be name `onepassword_facts`. While I'm sure this works, there are many places in Python where using a digit at the beginning is unsupported, and we are asking for trouble with this name.

File: lib/ansible/modules/identity/1password_facts.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+#
+# (c) 2018, Ryan Conway (@rylon)
+# (c) 2018, Scott Buchanan <sbuchanan@ri.pn> (onepassword.py used as starting point)
+# (c) 2018, Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+
+# Ansible boilerplate: https://docs.ansible.com/ansible/2.6/dev_guide/testing/sanity/boilerplate.html
```
Comment (samdoran) [inline lib/ansible/modules/identity/1password_facts.py]: This line is not needed.

File: lib/ansible/modules/identity/1password_facts.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+
+DOCUMENTATION = '''
+module: 1password_facts
+author:
+    - Ryan Conway (@rylon)
+version_added: "2.7"
+requirements:
+    - C(op) 1Password command line utility (v0.5.1). See U(https://support.1password.com/command-line/)
+notes:
+    - "Based on the `onepassword` lookup plugin by Scott Buchanan <sbuchanan@ri.pn>."
```
Comment (samdoran) [inline lib/ansible/modules/identity/1password_facts.py]: Inline code formatting is done by wrapping with `C()`.

File: lib/ansible/modules/identity/1password_facts.py [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+short_description: Fetch facts from 1Password items
+description:
+    - M(1password_facts) wraps the C(op) command line utility to fetch data about one or more 1password items and return as Ansible facts.
+    - A fatal error occurs if any of the items being searched for can not be found.
+    - Recommend using with the I(no_log) option to avoid logging the values of the secrets being retrieved.
```
Comment (samdoran) [inline lib/ansible/modules/identity/1password_facts.py]: `I(no_log)` => `C(no_log)`

**Review by samdoran — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/identity/1password_facts.py (hunk 1 above)
Comment (Rylon) [reply lib/ansible/modules/identity/1password_facts.py]: Yeah good point, I'll rename it to be on the safe side.

File: lib/ansible/modules/identity/1password_facts.py (hunk 2 above)
Comment (Rylon) [reply lib/ansible/modules/identity/1password_facts.py]: Ah yes, that was for me, I will remove it.

**Comment by Rylon** [comment]
@samdoran thanks for your review, I've made the requested changes 👍

**Review by samdoran — APPROVED** [review]

**Comment by samdoran** [comment]
I'm going to merge this to make the cutoff and it is mostly working. I hit a few items it fails to parse properly, so the parsing section needs some work (it's really tough).

I would also like to unify the terms used between this and the lookup plugins.

I like how you handled the auto-login. I just added similar capability to the lookup plugins. Again, we should unify this behavior before the final 2.7 release.

**Comment by Rylon** [comment]
Awesome thanks @samdoran to be clear is there anything you would like me to do as part of unifying, between this and the 1Password lookup plugin? Happy to be involved and help where I can for the 2.7 release.

**Comment by samdoran** [comment]
I mainly want to collaborate and come up with common terms and interface for handling the various sign in scenarios since we both added that capability.

Initial sign-in requires:
- `signinaddress`
- `emailaddress`
- `secretkey`
- `password`

Re-authenticated sign-in requires:
- `subdomain` (not _technically_ required, but I would like the module/lookup to require this)
- `password`

Already authenticated just reuses the existing session token in the environment.

You handled the initial sign-in while I only handled re-authented sign-in. I would mainly like to unify these methods and see if there is way to avoid asking for `signinaddress`, `emailaddress`, and `secretkey` every time if the system has already been signed in once. And possibly not require the master password if a token exists in the environment (the lookup originally behaved this way, requiring the user to handle the sign in outside of Ansible).

There credentials are _incredibly sensitive_ and I want to handle them as little as possible.

**Comment by samdoran** [comment]
Regarding terms, most of what we used is in alignment. I used `vault_password` where you used `masterpassword`. I propose we standardize on the following terms:

- `signin_url`: URL of the 1Password account, e.g., `example.1password.com`
- `username`: Email address used to authenticate to 1Password
- `master_password`: The master password to the vault
- `secret_key`: The secret key needed for initial sign-in

Thinking through this, we could ditch `subdomain` in favor of asking for `sigin_url`, or only require `subdomain` if `signin_url` isn't provided since we can pull `subdomain` from the `signin_url`.

**Comment by Rylon** [comment]
Yeah, that makes sense. The reason I avoided differentiating between initial sign-in and re-authenticated sign-in was to keep the code logic as simple as possible. This way we avoid needing to track any auth state, and avoid having to handle error cases such as an attempt to re-authenticate where the subdomain provided doesn't match an existing authenticated account.

If `auto_login` is not specified at all, the module should already behave the same way as the lookup plugin, requiring the user to handle signing outside of Ansible, but if it is specified, then this does not work. The reason for that was to make it a conscious choice by the user and to avoid any "magic" behaviour.

Changing `subdomain` to `signin_url` seems like a sensible idea, I think it is better to be explicit and makes it easier to update if 1Password change anything. The rest of the proposed terms look good 👍

**Pull request merged**